sw/device/tests/rv_core_ibex_address_translation_test.c

```diff
@@ -0,0 +1,295 @@
+    CHECK_DIF_OK(dif_rv_core_ibex_configure_addr_translation(
+        ibex_core, slot, kBuses[idx], mapping));
+  };
+}
+
+/**
+ * Enables IBus and DBus address translation for the given slot.
+ *
+ * @param ibex_core A handle to the ibex core.
+ * @param slot The slot to be used enabled.
+ */
+void enable_slot(dif_rv_core_ibex_t *ibex_core,
+                 dif_rv_core_ibex_addr_translation_slot_t slot) {
+  for (size_t idx = 0; idx < 2; ++idx) {
+    CHECK_DIF_OK(
+        dif_rv_core_ibex_enable_addr_translation(ibex_core, slot, kBuses[idx]));
+  };
+}
+
+bool test_main(void) {
+  // Calculate 2-byte aligned addresses to put a copy of the toy functions in.
+  uintptr_t flash_mem_end_addr = TOP_EARLGREY_FLASH_CTRL_MEM_BASE_ADDR +
+                                 TOP_EARLGREY_FLASH_CTRL_MEM_SIZE_BYTES;
+  uintptr_t offset = 0x1000;
+  uintptr_t make_lower_case_addr = flash_mem_end_addr - offset;
+  uintptr_t get_name_addr = flash_mem_end_addr - (2 * offset);
+
+  // Initiate the flash controller.
+  dif_flash_ctrl_state_t flash_ctrl = init_flash();
+
+  // Move the toy functions to their aligned addresses.
+  CHECK(flash_ctrl_testutils_erase_and_write_page(
+      /*flash_state=*/&flash_ctrl,
+      /*byte_address=*/(uint32_t)make_lower_case_addr,
+      /*partition_id=*/0,
+      /*data=*/(uint32_t *)make_lower_case,
+      /*partition_type=*/kDifFlashCtrlPartitionTypeData,
+      /*word_count=*/kMakeLowerCaseFnSize / 4));
+
+  CHECK(flash_ctrl_testutils_erase_and_write_page(
+      /*flash_state=*/&flash_ctrl,
+      /*byte_address=*/(uint32_t)get_name_addr,
+      /*partition_id=*/0,
+      /*data=*/(uint32_t *)get_name,
+      /*partition_type=*/kDifFlashCtrlPartitionTypeData,
+      /*word_count=*/kGetNameFnSize / 4));
+
+  char test_str[] = TEST_STR;
+  ((str_fn_t)make_lower_case_addr)(test_str);
+  CHECK_STR_EQ(test_str, EXPECTED_RESULT_MAKE_LOWER_CASE);
+
+  ((str_fn_t)get_name_addr)(test_str);
+  CHECK_STR_EQ(test_str, EXPECTED_RESULT_GET_NAME);
+
+  // The memory address to which the functions will be mapped.
+  const uintptr_t kRemapAddr = 0xA0000000;
+
+  // Create translation descriptions.
+  dif_rv_core_ibex_addr_translation_mapping_t make_lower_case_mapping = {
+      .matching_addr = kRemapAddr,
+      .remap_addr = make_lower_case_addr,
+      .size = next_power_of_two(kMakeLowerCaseFnSize),
+  };
+  dif_rv_core_ibex_addr_translation_mapping_t get_name_mapping = {
+      .matching_addr = kRemapAddr,
+      .remap_addr = get_name_addr,
+      .size = next_power_of_two(kGetNameFnSize),
+  };
+
+  // Get ibex core handle.
+  dif_rv_core_ibex_t ibex_core;
+  CHECK_DIF_OK(dif_rv_core_ibex_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
+      &ibex_core));
+
+  // Map virtual address space to make_lower_case() using slot 1.
+  map_to_slot(&ibex_core, kSlots[1], make_lower_case_mapping);
+
+  // Enable address translation slot 1.
+  enable_slot(&ibex_core, kSlots[1]);
```

Comment:
Should you also disable slot 0?

Reply:
This is a good point. [Reset value is zero](https://docs.opentitan.org/hw/ip/rv_core_ibex/doc/index.html#Reg_ibus_addr_en_0), so I'm happy for the test to fail if slot zero is enabled at start up.